File: ds&algo/binarySearch.js
const binarySearch = (array, left, right, number) => {
	while (left <= right) {
		let mid = parseInt((left + right) / 2);
		if (array[mid] === number) {
			return mid;
		}
		if (array[mid] < number) {
			left = mid + 1;
		} else {
			right = mid - 1;
		}
	}
	return -1;
};

const binarySearchRecursive = (array, left, right, number) => {
	while (left <= right) {
		let mid = parseInt((left + right) / 2);
		if (array[mid] === number) {
			return mid;
		}
		if (array[mid] < number) {
			return binarySearchRecursive(array, mid + 1, right, number);
		} else {
			return binarySearchRecursive(array, left, mid - 1, number);
		}
	}
	return -1;
};

function main() {
	const array = [2, 10, 11, 15, 17, 28, 31, 33, 39, 46];
	const number = 33;
	const length = array.length;
	const result = binarySearch(array, 0, length - 1, number);
	result !== -1 ? console.log(`${number} found at index ${result}`) : console.log(`${number} not found`);
}

main();

File: ds&algo@cpp/binarySearch.cpp
#include <iostream>
using namespace std;

int binarySearch(int arr[], int length, int key)
{
    int left = 0;
    int right = length - 1;
    int mid;

    while (left <= right)
    {
        mid = (left + right) / 2;
        if (arr[mid] == key)
        {
            return mid;
        }
        if (arr[mid] > key)
        {
            right = mid - 1;
        }
        else
        {
            left = mid + 1;
        }
    }
    return -1;
}

int binarySearchRecursive(int arr[], int left, int right, int key)
{
    if (left <= right)
    {
        int mid = (left + right) / 2;
        
        if (arr[mid] == key)
        {
            return mid;
        }
        if (arr[mid] > key)
        {
            return binarySearchRecursive(arr, left, mid - 1, key);
        }
        else
        {
            return binarySearchRecursive(arr, mid + 1, right, key);
        }
    }

    return -1;
}

int main()
{
    int arr[] = {21, 35, 48, 56, 71, 84, 95};
    int length = 7;
    int key = 84;
    int left = 0;
    int right = length - 1;
    // int result = binarySearch(arr, length, key);
    int result = binarySearchRecursive(arr, left, right, key);
    cout << "The " << key << " Found in index " << result << endl;
    return 0;
}